smart_open/bytebuffer.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+"""Implements a byte buffer that allows callers to read data with minimal
+copying, and has a fast __len__ method. The buffer is parametrized by its
+chunk_size, which is the number of bytes that it will read in from the supplied
+reader or iterable when the buffer is being filled. As primary use case for this
+buffer is to amortize the overhead costs of transferring data over the network
+(rather than capping memory consumption), it leads to more predictable
+performance to always read the same amount of bytes each time the buffer is
+filled, hence the chunk_size parameter instead of some fixed capacity.
+
+The bytes are stored in a bytestring, and previously-read bytes are freed when
+the buffer is next filled (by slicing the bytestring into a smaller copy)."""
+
+class ByteBuffer(object):
```

Comment:
Please add a docstring for this class. I think most of the above text (module docstring) is really the class' docstring.

Are you familiar with doctest? If you are, please add some code examples of using your new class to the class or module docstring, e.g.

```python
>>> buf = ByteBuffer()
>>> buf.fill('hello world')
>>> len(buf)
11
>>> buf.read()
hello world
>>> # and so on
```

Also, PEP8 requires two blank lines before class declarations.

Reply:
I've never used doctest before, but it seemed simple enough, so I gave it a shot. However, I ran into problems with making the same doctest work in both Python 2 and 3. If you're interested in the details, I pushed up [a commit with the doctest I tried to use and a commit message describing the problem to another branch on my fork of smart_open](https://github.com/aperiodic/smart_open/commit/9ab06331e880988dca7b33f7f89a1ecce01f53f4).

Reply:
The class docstring is addressed in https://github.com/RaRe-Technologies/smart_open/pull/284/commits/02ca639b326b9147b4d8d99b6fb8eca6e3acf3b8

Reply:
I've added the doctest (with a note that it only passes in Python 3) in https://github.com/RaRe-Technologies/smart_open/pull/284/commits/c604a78b56ea3214dcc8488c68d6fcfdd1410613.